Objects/typeobject.c

```diff
@@ -287,32 +290,51 @@ type_mro_modified(PyTypeObject *type, PyObject *bases) {
 
        Unset HAVE_VERSION_TAG and VALID_VERSION_TAG if the type
        has a custom MRO that includes a type which is not officially
-       super type.
+       super type, or if the type implements its own mro() method.
 
        Called from mro_internal, which will subsequently be called on
        each subclass when their mro is recursively updated.
      */
     Py_ssize_t i, n;
     int clear = 0;
+    int custom = (Py_TYPE(type) != &PyType_Type);
 
     if (!PyType_HasFeature(type, Py_TPFLAGS_HAVE_VERSION_TAG))
         return;
 
-    n = PyTuple_GET_SIZE(bases);
-    for (i = 0; i < n; i++) {
-        PyObject *b = PyTuple_GET_ITEM(bases, i);
-        PyTypeObject *cls;
-
-        assert(PyType_Check(b));
-        cls = (PyTypeObject *)b;
-
-        if (!PyType_HasFeature(cls, Py_TPFLAGS_HAVE_VERSION_TAG) ||
-            !PyType_IsSubtype(type, cls)) {
+    if (custom) {
+        _Py_IDENTIFIER(mro);
+        int unbound;
+        PyObject *mro_meth = lookup_maybe_method(
+            (PyObject *)type, &PyId_mro, &unbound);
```

Comment:
`mro_meth` needs to be checked for NULL before `lookup_maybe_method` is called again.

Reply:
OK :-] is using a goto legit here? It'll avoid a bunch of `if (clear == 0)` (see current impl).